examples/authentication/client/src/main.rs

```diff
@@ -0,0 +1,194 @@
+    // TODO(#886): Consider retrieving scope from server.
+    let scope = "openid email";
+    let redirect_uri = format!("http://{}", opt.redirect_address);
+    // TODO(#886): Retrieve state and code challenge from server and add to request.
+    auth_endpoint
+        .query_pairs_mut()
+        .append_pair("scope", scope)
+        .append_pair("response_type", "code")
+        .append_pair("redirect_uri", &redirect_uri)
+        .append_pair("client_id", &opt.client_id);
+    auth_endpoint.to_string()
+}
+
+struct ResultSender {
+    sender: Option<Sender<()>>,
+    result: Option<Result<String, String>>,
+}
+
+impl ResultSender {
+    fn notify(&mut self) {
+        if self.sender.is_some() {
+            self.sender.take().unwrap().send(()).unwrap();
+        }
+    }
+}
+
+// Handles the redirects to extract code from the query string.
+struct RedirectHandler {
+    result_sender: Arc<Mutex<ResultSender>>,
+}
+
+impl Service<Request<Body>> for RedirectHandler {
+    type Response = Response<Body>;
+    type Error = hyper::Error;
+    type Future = future::Ready<Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Ok(()).into()
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        if let Some(query) = request.uri().query() {
+            let lookup: HashMap<_, _> = form_urlencoded::parse(query.as_bytes()).collect();
+            let result_sender = &mut self.result_sender.lock().unwrap();
+            if lookup.contains_key("code") {
+                let code = lookup.get("code").unwrap().to_string();
+                info!("Auth code: {:?}", code);
+                result_sender.result = Some(Ok(code));
+                result_sender.notify();
+                future::ok(Response::new(Body::from("Success!")))
+            } else if lookup.contains_key("error") {
+                let error = lookup.get("error").unwrap().to_string();
+                warn!("Error: {:?}", error);
+                result_sender.result = Some(Ok(error));
+                result_sender.notify();
+                future::ok(
+                    Response::builder()
+                        .status(StatusCode::UNAUTHORIZED)
+                        .body(Body::from("Authentication failed!"))
+                        .unwrap(),
+                )
+            } else {
+                future::ok(
+                    Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::from("Invalid query string!"))
+                        .unwrap(),
+                )
+            }
+        } else {
+            future::ok(
+                Response::builder()
+                    .status(StatusCode::BAD_REQUEST)
+                    .body(Body::from("No query string!"))
+                    .unwrap(),
+            )
+        }
+    }
+}
+
+// Produces instances of the redirect handler service.
```

Comment:
Please add some more documentation, either here or on `RedirectHandler`, the code seems fine but it takes a while to figure out what's talking to what, so perhaps a high level description may help.

Reply:
Done.